components/gitpod-protocol/src/util/generate-workspace-id.ts

```diff
@@ -5,8 +5,25 @@
  */
 import randomNumber = require("random-number-csprng");
 
-export async function generateWorkspaceID(): Promise<string> {
-    return (await random(colors))+'-'+(await random(animals))+'-'+(await random(characters, 8));
+export async function generateWorkspaceID(firstSegment?: string, secondSegment?: string): Promise<string> {
+    const firstSeg = clean(firstSegment) || await random(colors);
+    const secSeg = clean(secondSegment, Math.min(16, 24-firstSeg.length)) || await random(animals);
+    return firstSeg+'-'+secSeg+'-'+(await random(characters, 8));
+}
+
+function clean(segment: string | undefined, maxChars: number = 16) {
+  if (segment) {
```

Comment:
Do I read this correct as:
```
if (!segment) return undefined;

const result = Array.from(segment)
  .filter(c => characters.indexOf(c) !== -1)
  .join('');
if (result.length >= 2) {
   return result.substring(0, maxChars);
}
return result;
```
:thinking:

Reply:
Ok, I rewrote it as suggested.